content/unroll-default-arguments.md

```diff
@@ -0,0 +1,933 @@
+---
+layout: sip
+permalink: /sips/:title.html
+stage: completed
+status: under-review
+title: SIP-61 - Unroll Default Arguments for Binary Compatibility
+---
+
+**By: Li Haoyi**
+
+## History
+
+| Date          | Version            |
+|---------------|--------------------|
+| Feb 14th 2024 | Initial Draft      |
+
+## Summary
+
+This SIP proposes an `@unroll` annotation lets you add additional parameters
+to method `def`s,`class` construtors, or `case class`es, without breaking binary 
+compatibility. `@unroll` works by generating "unrolled" or "telescoping" forwarders:
```

Comment:
Regarding the naming, I don't mind very much either way. But _compatibility_ is a concept which doesn't exist within one class, but only as a relationship between that class in two versions of that code base. So maybe it would be better to name the annotation only after the mechanism (which incidentally helps with compatibility). The two Scala mechanisms at play here are "Default Parameters" and "Overloading". If we use them, the name could be
 * `@overloadDefaultParams`
 * `@defaultParamsOverload`
 * `@overloadDefaultParameters`
 * `@defaultParametersOverload`

But I have to admit that it's longer than `@unroll`.

Reply:
If you can annotate the whole thing the annotation can be kept on its own line and then the length of the name is not that big of a deal.

Reply:
I agree that compatibility intent is difficult after thinking for a while. 

I think I like `@overloadDefaultParam` best as the name of a "whole-thing-annotation".

Reply:
Yes if we annotate the whole method the length does not matter, but if we annotation individual parameters the length does matter.

One option is to specify individual parameter names in a method-level annotation `@unroll("param1", "param2") def foo...` , but that involves duplication of the parameter names and introduces an extra degree of freedom for people to make mistakes.

With the current approach of annotating the parameters themselves, it is important to keep the annotation short. And I'll say again that any one confused about `@unroll` is one click away from a detailed explanation in the Scaladoc, so I don't think we need to have a long and verbose name. 

It's not like people unfamiliar can look at `.pipe` or `.tap` and immediately know what it means either, but because the detailed Scaladoc is one click away the initial confusion is easy to overcome and being concise can take priority over being verbose and spelling out exactly what the functionality does in a long and verbose identifier. I believe the same logic applies to `@unroll`

Reply:
Thanks for taking the time to explain all rationale! As usual, you have good answers to everything! 

OK, I can live with `@unroll` on individual params. But we could still *also* have an annotation on the whole thing and the param to that annotation could be from where the unrolling starts, like in:
```scala
@unrollDefaultFrom("y")
case class Thing(x: Int, y: Int = 0, z: Int= 0)
``` 
as a more readable (when applicable) alternative to this:
```scala
case class Thing(x: Int, @unroll y: Int = 0, @unroll z: Int= 0)
```

Reply:
>  But we could still also have an annotation on the whole thing and the param to that annotation could be from where the unrolling starts, like in:

I would personally be against this. Having two ways to spell the same thing seems worse than having either one alone. When there's backwards compatibility or migration concerns, having some duplication seems fine, but since we have a chance to specify this from scratch we should really just pick one.

It basically comes down to the age-old tradeoff between "intrusive annotation" or "lookaside table", and the current proposal of "intrusive annotation" is a decent spot in a solution space with no "correct" answer.

Reply:
Well, the proposal already has two ways o doing things: `@unroll` and `@unrollAll`. But I agree that it's a matter of taste. I think I prefer that the unrollAll variant is placed on the whole thing with a param from where it starts. That way we both get the cake and can eat it, so to speak.